src/pool/spawn.rs

```diff
@@ -1,56 +1,269 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
-use crate::queue::{TaskCell, TaskInjector};
+//! This module implements how task are pushed and polled. Threads are
+//! woken up when new tasks arrived and go to sleep when there are no
+//! tasks waiting to be handled.
+
+use crate::pool::SchedConfig;
+use crate::queue::{LocalQueue, Pop, TaskCell, TaskInjector};
+use parking_lot_core::{ParkResult, ParkToken, UnparkToken};
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 
-/// Allows spawning a task to the thread pool from a different thread.
-pub trait RemoteSpawn: Sync + Send {
-    /// The task it can spawn.
-    type TaskCell;
+/// An usize is used to trace the threads that are working actively.
+/// To save additional memory and atomic operation, the number and
+/// shutdown hint are merged into one number in the following format
+/// ```text
+/// 0...00
+/// ^    ^
+/// |    The least significant bit indicates whether the queue is shutting down.
+/// Bits represent the thread count
+/// ```
+const SHUTDOWN_BIT: usize = 1;
+const WORKER_COUNT_SHIFT: usize = 1;
+const WORKER_COUNT_BASE: usize = 2;
+
+/// Checks if shutdown bit is set.
+pub fn is_shutdown(cnt: usize) -> bool {
+    cnt & SHUTDOWN_BIT == SHUTDOWN_BIT
+}
 
-    /// Spawns a task into the thread pool.
-    fn spawn(&self, task_cell: Self::TaskCell);
+/// The core of queues.
+///
+/// Every thread pool instance should have one and only `QueueCore`. It's
+/// saved in an `Arc` and shared between all worker threads and remote handles.
+pub(crate) struct QueueCore<T> {
+    global_queue: TaskInjector<T>,
+    active_workers: AtomicUsize,
+    config: SchedConfig,
 }
 
-/// Allows spawning a task inside the thread pool.
-pub trait LocalSpawn {
-    /// The task it can spawn.
-    type TaskCell;
-    /// The remote handle that can be used in other threads.
-    type Remote: RemoteSpawn<TaskCell = Self::TaskCell>;
+impl<T> QueueCore<T> {
+    pub fn new(global_queue: TaskInjector<T>, config: SchedConfig) -> QueueCore<T> {
+        QueueCore {
+            global_queue,
+            active_workers: AtomicUsize::new(config.max_thread_count << WORKER_COUNT_SHIFT),
+            config,
+        }
+    }
 
-    /// Spawns a task into the thread pool.
-    fn spawn(&mut self, task: Self::TaskCell);
+    /// Ensures there are enough workers to handle pending tasks.
+    ///
+    /// If the method is going to wake up any threads, source is used to trace who triggers
+    /// the action.
+    pub fn ensure_workers(&self, source: usize) {
+        let cnt = self.active_workers.load(Ordering::Relaxed);
+        if (cnt >> WORKER_COUNT_SHIFT) >= self.config.max_thread_count || is_shutdown(cnt) {
```

Comment:
how about defining an independent function to compute the thread count? which is similar with is_shutdown